python/ql/src/Security/CWE-079/Jinja2WithoutEscaping.ql

```diff
@@ -25,25 +27,25 @@ import python
  * safe1_tmpl = Template('Hello {{ name }}!', autoescape=True)
  */
 
-ClassValue jinja2EnvironmentOrTemplate() {
-  result = Value::named("jinja2.Environment")
+private API::Node jinja2EnvironmentOrTemplate() {
+  result = API::moduleImport("jinja2").getMember("Environment")
   or
-  result = Value::named("jinja2.Template")
+  result = API::moduleImport("jinja2").getMember("Template")
 }
 
-ControlFlowNode getAutoEscapeParameter(CallNode call) { result = call.getArgByName("autoescape") }
+DataFlow::Node getAutoEscapeParameter(DataFlow::CallCfgNode call) {
+  result = call.getArgByName("autoescape")
+}
 
-from CallNode call
+from DataFlow::CallCfgNode call
 where
-  call.getFunction().pointsTo(jinja2EnvironmentOrTemplate()) and
-  not exists(call.getNode().getStarargs()) and
-  not exists(call.getNode().getKwargs()) and
+  call = jinja2EnvironmentOrTemplate().getACall() and
+  not exists(call.asCfgNode().(CallNode).getNode().getStarargs()) and
+  not exists(call.asCfgNode().(CallNode).getNode().getKwargs()) and
   (
     not exists(getAutoEscapeParameter(call))
     or
-    exists(Value isFalse |
-      getAutoEscapeParameter(call).pointsTo(isFalse) and
-      isFalse.getDefiniteBooleanValue() = false
-    )
+    any(DataFlow::LocalSourceNode n | n.asExpr().(ImmutableLiteral).booleanValue() = false)
+        .flowsTo(getAutoEscapeParameter(call))
```

Comment:
if `call` was an `API::CallNode` you could use the following code

```suggestion
    call.getKeywordParameter("autoescape").getAValueReachingRhs().asExpr().(ImmutableLiteral).booleanValue() = false
```

to also get global flow -- that's how I've been tracking boolean values in other library modeling, so I think doing it here would also make sense :blush: 

doing this would require inlining `getAutoEscapeParameter`, which I think is fine

Reply:
I just tried this, and it introduced a false positive in
```python
def checked(cond=False):
    if cond:
        e = Environment(autoescape=cond) # GOOD (but now has an alert)
```
So I'm inclined _not_ to do it at the present time. I think this may have to do with flow coming out of default values not being _quite_ right.

I'll switch `DataFlow::CallCfgNode` anyway, though, as that'll set us up better for the future.

Reply:
> I think this may have to do with flow coming out of default values not being _quite_ right.

Is it not that the `if` does not act as a barrier, since there is really not a concept of value involved in type tracking? It seems correct to take the default value into account..

Reply:
Ah, you're right. I think I had somehow inverted the logic in my head, thinking it was `cond=True` rather than `cond=False`, and that we were seeing the alert due to missing flow (to a call that was in fact safe).

This makes me wonder, though, if we couldn't incorporate these barriers into our local flow (and hence type tracking). Consider e.g. splitting the local flow relation into two forms, one for truthy values, and one for falsy values. We could then only allow flow into the body of an `if` like the above if it is of the correct truthyness. (The general local flow relation would then be the union of the truthy and falsy relations.)

It's probably not worth the effort, but it would be an interesting experiment.

Reply:
> to also get global flow -- that's how I've been tracking boolean values in other library modeling, so I think doing it here would also make sense 😊

I am generally fine with rewriting `pointsTo` into `getAValuereachingRhs` as a quick way to convert our queries. I will submit, though, that this does not really get global flow, but a different global property, and in cases where global flow is important, we should use global data flow.